File: cpp/image-acquisition/save-bmp-on-trigger/src/save-bmp-on-trigger.cpp
#include <ic4/ic4.h>

#include <console-helper.h>

#include <iostream>


// Define QueueSinkListener-derived class that saves all received frames in bitmap files
class SaveAsBmpListener : public ic4::QueueSinkListener
{
private:
	std::string path_base_;
	int counter_;

public:
	SaveAsBmpListener(std::string path_base)
		: path_base_(std::move(path_base))
		, counter_(0)
	{
	}

	// Inherited via QueueSinkListener, called when there are frames available in the sink's output queue
	void framesQueued(ic4::QueueSink& sink) override
	{
		ic4::Error err;

		while (true)
		{
			// Remove a buffer from the sink's output queue
			auto buffer = sink.popOutputBuffer(err);
			if (buffer == nullptr)
			{
				// No more buffers available, return
				return;
			}

			// Generate a file name for the bitmap file
			auto file_name = path_base_ + std::to_string(counter_) + ".bmp";

			// Save the image buffer in the bitmap file
			if (!ic4::imageBufferSaveAsBitmap(*buffer, file_name, {}, err))
			{
				std::cerr << "Failed save buffer: " << err.message() << std::endl;
			}
			else
			{
				std::cout << "Saved image " << file_name << std::endl;

				counter_ += 1;
			}
		}
	}
};

int main()
{
	ic4::initLibrary();
	std::atexit(ic4::exitLibrary);

	// Let the user select a device
	auto device_list = ic4::DeviceEnum::enumDevices();
	auto it = ic4_examples::console::select_from_list(device_list);
	if (it == device_list.end())
	{
		return -1;
	}

	// Open the selected device
	ic4::Error err;
	ic4::Grabber grabber;
	if (!grabber.deviceOpen(*it, err))
	{
		std::cerr << "Failed to open device: " << err.message() << std::endl;
		return -2;
	}

	auto map = grabber.devicePropertyMap(err);
	if (err.isError())
	{
		std::cerr << "Failed to query device property map: " << err.message() << std::endl;
		return -3;
	}

	// Reset all device settings to default
	// Not all devices support this, so ignore possible errors
	map.setValue(ic4::PropId::UserSetSelector, "Default", ic4::Error::Ignore());
	map.executeCommand(ic4::PropId::UserSetLoad, ic4::Error::Ignore());

	// Select FrameStart trigger (for cameras that support this)
	map.setValue(ic4::PropId::TriggerSelector, "FrameStart", ic4::Error::Ignore());

	// Enable trigger mode
	if (!map.setValue(ic4::PropId::TriggerMode, "On", err))
	{
		std::cerr << "Failed to enable trigger mode: " << err.message() << std::endl;
		return -4;
	}
	
	// Create an instance of the listener type defined above, specifying a partial file name
	std::string path_base = "./test_image";
	SaveAsBmpListener listener(path_base);

	// Create a QueueSink to capture all images arriving from the video capture device
	auto sink = ic4::QueueSink::create(listener, err);
	if (!sink)
	{
		std::cerr << "Failed to create sink: " << err.message() << std::endl;
		return -3;
	}

	// Start the video stream into the sink
	if (!grabber.streamSetup(sink, ic4::StreamSetupOption::AcquisitionStart, err))
	{
		std::cerr << "Failed to setup stream: " << err.message() << std::endl;
		return -4;
	}

	std::cout << "Stream started." << std::endl;
	std::cout << "Waiting for triggers" << std::endl;
	std::cout << "All images will be saved as " << path_base << "*.bmp" << std::endl;
	std::cout << std::endl;

	std::cout << "Input hardware triggers, or press ENTER to issue a software trigger" << std::endl;
	std::cout << "Press q + ENTER to quit" << std::endl;

	while( true )
	{
		int ch = std::getchar();
		if (ch == 'q')
			break;

		// Execute software trigger
		if (!map.executeCommand(ic4::PropId::TriggerSoftware, err))
		{
			std::cerr << "Failed to perform software trigger: " << err.message() << std::endl;
			continue;
		}
	}

	// We have to call streamStop before exiting the function, because we have the listener defined as a stack variable.
	grabber.streamStop();
	grabber.deviceClose();

	return 0;
}

File: python/image-acquisition/save-bmp-on-trigger/save-bmp-on-trigger.py

import imagingcontrol4 as ic4


def example_save_bmp_on_trigger():
    # Let the user select one of the connected cameras
    device_list = ic4.DeviceEnum.devices()
    for i, dev in enumerate(device_list):
        print(f"[{i}] {dev.model_name} ({dev.serial}) [{dev.interface.display_name}]")
    print(f"Select device [0..{len(device_list) - 1}]: ", end="")
    selected_index = int(input())
    dev_info = device_list[selected_index]

    # Open the selected device in a new Grabber
    grabber = ic4.Grabber(dev_info)
    map = grabber.device_property_map

    # Reset all device settings to default
    # Not all devices support this, so ignore possible errors
    map.try_set_value(ic4.PropId.USER_SET_SELECTOR, "Default")
    map.try_set_value(ic4.PropId.USER_SET_LOAD, 1)

    # Select FrameStart trigger (for cameras that support this)
    map.try_set_value(ic4.PropId.TRIGGER_SELECTOR, "FrameStart")

    # Enable trigger mode
    map.set_value(ic4.PropId.TRIGGER_MODE, "On")

    # Define a listener class to receive queue sink notifications
    class Listener(ic4.QueueSinkListener):
        def __init__(self, path_base: str):
            self.path_base = path_base
            self.counter = 0

        def sink_connected(self, sink: ic4.QueueSink, image_type: ic4.ImageType, min_buffers_required: int) -> bool:
            # No need to configure anything, just accept the connection
            return True
        
        def frames_queued(self, sink: ic4.QueueSink):
            # Get the queued image buffer
            buffer = sink.pop_output_buffer()

            # Save the image buffer's contents in a BMP file
            file_name = f"{self.path_base}{self.counter}.bmp"
            buffer.save_as_bmp(file_name)

            print(f"Saved image {file_name}")

            self.counter = self.counter + 1

    # Create an instance of the listener type defined above, specifying a partial file name
    path_base = "./test_image"
    listener = Listener(path_base)

    # Create a QueueSink to capture all images arriving from the video capture device
    sink = ic4.QueueSink(listener)

    # Start the video stream into the sink
    grabber.stream_setup(sink)

    print("Stream started.")
    print("Waiting for triggers")
    print(f"All images will be saved as {path_base}*.bmp")
    print()
    print("Input hardware triggers, or press ENTER to issue a software trigger")
    print("Press q + ENTER to quit")

    while input() != 'q':
        # Execute software trigger
        map.execute_command(ic4.PropId.TRIGGER_SOFTWARE)

    # We have to call streamStop before exiting the function, to make sure the listener object is not destroyed before the stream is stopped
    grabber.stream_stop()

    map.set_value(ic4.PropId.TRIGGER_MODE, "Off")

    grabber.device_close()

if __name__ == "__main__":
    with ic4.Library.init_context(api_log_level=ic4.LogLevel.INFO, log_targets=ic4.LogTarget.STDERR):

        example_save_bmp_on_trigger()
